src/game.rs

```diff
@@ -21,90 +21,238 @@ pub enum ChessError {
+                let mv = Move::new(from, to);
+                self.is_move_valid(&mv)?;
 
-            self.is_move_valid(&mv)?;
-            self.state.move_piece(&mv);
-            self.add_moves();
-            self.remove_self_checks();
-            self.update_status();
+                let mut next_state = self.clone_current_state();
+                next_state.move_piece(&mv);
+                self.history.add_info(next_state, mv);
+                self.update();
 
-            println!("{} : {}", piece, mv);
+                println!("{} : {}", piece, mv);
+            }
         }
         Ok(())
     }
 
+    fn update(&mut self) {
+        self.add_moves();
+        self.remove_self_checks();
+        self.update_status()
+    }
+
     fn update_status(&mut self) {
-        let mut next_turn = self.clone();
-        next_turn.state.player = !next_turn.state.player;
-        next_turn.add_moves();
-        if next_turn.has_check() {
-            if self.valid_moves.is_empty() {
-                self.status = GameStatus::Checkmate;
-            } else {
-                self.status = GameStatus::Check;
-            }
-        } else if self.valid_moves.is_empty() {
-            self.status = GameStatus::Stalemate;
+        if self.history.is_replaying() {
+            self.status.update(GameStatus::Replay);
+            return;
+        }
+        let king_is_under_attack = self.is_king_under_attack();
+        let valid_moves_is_empty = self.valid_moves.is_empty();
+
+        if !king_is_under_attack && valid_moves_is_empty {
+            self.status.update(GameStatus::Stalemate)
+        } else if king_is_under_attack && valid_moves_is_empty {
+            self.status.update(GameStatus::Checkmate)
+        } else if king_is_under_attack {
+            self.status.update(GameStatus::Check)
         } else {
-            self.status = GameStatus::Ongoing;
+            self.status.update(GameStatus::Ongoing)
         }
     }
 
-    fn has_check(&self) -> bool {
+    fn is_attacking_king(&self) -> bool {
         self.valid_moves
             .iter()
-            .any(|m| self.get_piece(&m.to) == Some(Piece::King(!self.state.player)))
+            .any(|mv| self.get_piece(&mv.to) == Some(Piece::King(!self.get_current_player())))
+    }
+
+    fn is_king_under_attack(&self) -> bool {
+        let mut enemy_board = self.clone_current_state();
+        enemy_board.player = !enemy_board.player;
+        Self::with_state(enemy_board).is_attacking_king()
     }
 
     fn remove_self_checks(&mut self) {
-        for mv in self.valid_moves.clone() {
-            let mut next_turn = self.clone();
-            next_turn.state.move_piece(&mv);
-            next_turn.add_moves();
-            if next_turn.has_check() {
-                self.valid_moves.remove(&mv);
-            }
-        }
+        let current_board = self.clone_current_state();
+        self.valid_moves.retain(|mv| {
+            let mut future_board = current_board.clone();
```

Comment:
It would be dope if we could just call `clone_current_state` in here and not have to create the `current_board` at all. Might not work though.

Reply:
doesn't work